In [1]:
import numpy as np
from scipy.integrate import quad
from scipy import integrate

# integrate.trapezoid
# integrate.romberg # deprecated use trapezoid

# Define integrand_cyl() function
def integrand_cyl(x, paramwave):
    # Extract inner can radius (`R1`) and zeta
    R1 = paramwave[0]
    zeta = paramwave[1]

    dval = np.sqrt(R1**2 - x**2)
    integrand = 1 - np.exp(-2 * dval/zeta)
    return integrand

# Define integral_cyl() function
def integral_cyl(xs, unit_cell_volume, pack_fraction, R1):
    zeta = unit_cell_volume/pack_fraction/xs # Note: packing efficiency? units?
    paramwave = [R1, zeta]
    int1, _ = quad(integrand_cyl, -R1, R1, args = (paramwave))
    result = 100 * int1/(2 * R1)
    return result

# Define integrand_ann() function
def integrand_ann(x, paramwave):
    # Extract inner can radius (`R1`), zeta, and outer can radius (`R2`)
    R1 = paramwave[0]
    zeta = paramwave[1]
    R2 = paramwave[2]

    # If absolute value of `x` is less than inner can radius, then 0
    if abs(x) < R1:
        return 0
    # If absolute value of `x` is greater than outer can radius, then 0
    elif abs(x) > R2:
        return 0
    
    dval = np.sqrt(R2**2 - x**2)
    integrand = 1 - np.exp(-2 * dval/zeta)
    return integrand

# Define integral_ann() function
def integral_ann(xs, unit_cell_volume, pack_fraction, R1, R2):
    zeta = unit_cell_volume/pack_fraction/xs
    paramwave = [R1, zeta, R2]
    
    int1, _ = quad(integrand_ann, -R2, R2, args = (paramwave))
    result = 100 * int1/(2*R2)
    return result          

In [2]:
integral_cyl(62.116,204.425,0.5,0.4)

9.070685056301812

In [3]:
integral_cyl(62.116,204.425,0.2,1)

9.070685056301805

In [4]:
integral_cyl(62.116,204.425,0.5,1)

21.04474243679236

In [6]:
integral_cyl(62.116,204.425,0.5,1)

21.04474243679236

•print cylin_integralV3(62.116,204.425,0.5,0.4)
  9.07064
•print cylin_integralV3(62.116,204.425,0.5,1)
  21.0446
•print cylin_integralV3(62.116,204.425,0.2,1)
  9.07064
print annular_integralV3(62.116,204.425,0.5,0.2,0.4)
  6.8933
•print annular_integralV3(62.116,204.425,0.5,0.1,0.4)
  8.53494
•print annular_integralV3(62.116,204.425,0.5,0.0,0.4)
  9.07064
 
Trying different types of numerical integration.
Trapezoidal integration (default).
•print cylin_integralV3(62.116,204.425,0.5,1)
  21.0446
 
Romberg integration.
•print cylin_integralV3(62.116,204.425,0.5,1)
  21.0427
 
Gaussian Quadrature integration.
•print cylin_integralV3(62.116,204.425,0.5,1)
  21.0447

# OLD CODE BELOW

In [ ]:
# Define intcyl3() function
def intcyl3(x, inner_radius_2, zeta):
    dval = np.sqrt(inner_radius_2**2 - x**2)
    return 1 - np.exp(-2 * dval / zeta)

# Define cylin_integralV3() function
def cylin_integralV3(depth, volume, pack_fraction, inner_radius):
    zeta = volume / (pack_fraction * depth)
    inner_radius_2 = inner_radius
    
    # Compute numerical integration of intcyl3() over the interval [-inner_radius, inner_radius]
    integral_result, _ = quad(intcyl3, -inner_radius, inner_radius, args = (inner_radius_2, zeta))

    int1 = 100 * integral_result
    
    return int1 / (2 * inner_radius)

# Find percent of incident beam that is scattered (assume no absorption)
percent_scatter_cyl = cylin_integralV3(depth = scatter_depth, volume = unit_cell_volume, pack_fraction = pack_fraction, inner_radius= can_inner_radius)

# Find percent of incident beam that is absorbed (assume no scattering)
percent_absorb_cyl = cylin_integralV3(depth = absorb_depth, volume = unit_cell_volume, pack_fraction = pack_fraction, inner_radius = can_inner_radius)
